### PR TITLE
docs/design: add proposal for restructure tests

### LINKED ISSUE #26022: Tracking issue for restructure tests
## Brief Motivation

This issue tracks the efforts to restructure TiDB tests, focusing on unit tests and integration tests, especially on separating `integration_test.go` as well as unifying test infrastructure in [testify](https://github.com/stretchr/testify).

For the detailed design, see also:

* [x] #26023

## Join force

We are currently on the first phase on the effort and focus on migrating tests to testify. If you want to participant the movement, you can directly comment under any open issues below without assignee to require assignment.

You can take a look at https://github.com/pingcap/tidb/pull/26375 as an example to do the migration.

## Note

**Before writing code, please take a look at PRs merged already and keep consistent code style.**

* parallel
  * As mentioned in #30692, we decided to run all tests in serial.
  * ~Keep test cases as parallel as possible.~
  * ~When some test cases should be run in serial, take use of go testing subtests (`t.Run(name string, f func(t *T))`) and parallel the parent test only. In this way, test cases in the same subtests set run in serial.~
  * ~https://github.com/pingcap/tidb/issues/26022#issuecomment-891656074~
  * ~When test cases should be run in serial at top level, move them into `xxx_serial_test.go` file.~
* Avoid using suite of testify as discussed https://github.com/pingcap/tidb/pull/26082#issuecomment-877230972 and https://github.com/pingcap/tidb/issues/27178#issuecomment-921842586
* Make use of testify checkers when the original code writes plain go testing cases.
* `c.Assert` is almost `require.Xxx`

## Migration

* [x] #26025
* [x] #27177
* [x] #27178
* [x] #27844
* [x] #27179
* [x] #27180
* [x] #26156
* [x] #26306
* [x] #26155
* [x] #26081
* [x] #26854
* [x] #26855
* [x] #28319
* [x] #26436
* [x] #26100
* [x] #26740
* [x] #26741
* [x] #26742
* [x] #26743
* [x] #28887
* [x] #26856
* [x] #26857
* [x] #26858
* [x] #26860
* [x] #26861
* [x] #26862
* [x] #26744
* [x] #26745
* [x] #26746
* [x] #26863
* [x] #26864
* [x] #27140
* [x] #27141
* [x] #27142
* [x] #27874
* [x] #33461
* [x] #33462
* [x] #33463
* [x] #27143
* [x] #27139
* [x] #26865
* [x] #26866
* [x] #26867
* [x] #26537
* [x] #26747
* [x] #26374
* [x] #26521
* [x] #33453

## Kits

* [x] #26106
* [x] #26107
* [x] #26108
* [x] #26289

## Reference

* testify suffers from parallelizing tests in suite https://github.com/stretchr/testify/issues/187

### LINKED ISSUE #26023: Create design doc for restructure tests
Previous discussion happened on TiDB Internals forum, see also https://internals.tidb.io/t/topic/200.

### PR DESCRIPTION
**Pull request by tisonkun**
### What problem does this PR solve?

Issue Number: close #26023

This pull requests brings the design document for restructure tests, focusing on unit tests and integration tests, especially on separating `integration_test.go` as well as unifying test infrastructure.

### Release note

- No release note

### CODE REVIEW

**Review by disksing — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

File: docs/design/2021-06-27-restructure-tests.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# Proposal: Restructure TiDB Tests
+
+- Author(s): [tison](http://github.com/tison)
```
Comment (LangInteger) [inline docs/design/2021-06-27-restructure-tests.md:3]: This should be `https://github.com/TisonKun`

Comment (xhebox) [reply docs/design/2021-06-27-restructure-tests.md:3]: /cc @tisonkun

Comment (tisonkun) [reply docs/design/2021-06-27-restructure-tests.md:3]: Oops. Let me create a followup.

Comment (tisonkun) [reply docs/design/2021-06-27-restructure-tests.md:3]: FYI https://github.com/pingcap/tidb/pull/26517

**Pull request merged**